File: src/286WallsAndGates/286WallsAndGates.cpp
#include <vector>
using namespace std;

class Solution {
public:
  void wallsAndGates(vector<vector<int>> &rooms) {
    int n = rooms.size();
    if (n == 0)
      return;
    int m = rooms[0].size();
    if (m == 0)
      return;
    static const int offset_x[] = {-1, 1, 0, 0};
    static const int offset_y[] = {0, 0, -1, 1};

    for (int j = 0; j < n; j++) {
      for (int i = 0; i < m; i++) {
        if (rooms[j][i] == 0) {
          queue<pair<int, int>> q;
          q.push(make_pair(j * m + i, 0));
          while (!q.empty()) {
            auto p = q.front();
            q.pop();
            int x = p.first % m, y = p.first / m, d = p.second;
            for (int k = 0; k < 4; k++) {
              int u = x + offset_x[k], v = y + offset_y[k];
              if (u >= 0 && u < m && v >= 0 && v < n && rooms[v][u] > d + 1) {
                rooms[v][u] = d + 1;
                q.push(make_pair(v * m + u, d + 1));
              }
            }
          }
        }
      }
    }
  }
};

int main() { return 0; }

File: src/286WallsAndGates/286WallsAndGates.rs
use std::collections::VecDeque;
impl Solution {
    pub fn walls_and_gates(rooms: &mut Vec<Vec<i32>>) {
        let m = rooms.len(); if m==0 {return;}
        let n = rooms[0].len(); if n==0 {return;}
        let mut queue = VecDeque::new();
        for j in 0..m {
            for i in 0..n {
                if rooms[j][i] == 0 {
                    queue.push_back((j, i));
                }
            }
        }
        let offset_x = vec![-1,1,0,0];
        let offset_y = vec![0,0,-1,1];
        while !queue.is_empty() {
            let (j, i) = queue.pop_front().unwrap();
            for d in 0..4 {
                let y = j as i32 + offset_y[d];
                let x = i as i32 + offset_x[d];
                if x>=0 && y>=0 && x<n as i32 && y<m as i32 && 
                   rooms[y as usize][x as usize ] == std::i32::MAX {
                    rooms[y as usize][x as usize ] = rooms[j][i]+1;
                    queue.push_back((y as usize, x as usize));
                }
            }
        }
    }
}
